PEP 733: An Evaluation of Python's Public C API (#3491)
python/peps: peps/pep-0733.rst

GC with Python's GC
[`Issue 19 <https://github.com/capi-workgroup/problems/issues/19>`__].

Inject foreign stack frames to the traceback
[`Issue 18 <https://github.com/capi-workgroup/problems/issues/18>`__].

Concrete strings that can be used in other languages
[`Issue 16 <https://github.com/capi-workgroup/problems/issues/16>`__].

References
==========

1. `Python/C API Reference Manual <https://docs.python.org/3/c-api/index.html>`__
2. `2023 Language Summit Blog Post: Three Talks on the C API <https://pyfound.blogspot.com/2023/05/the-python-language-summit-2023-three.html>`__
3. `capi-workgroup on GitHub <https://github.com/capi-workgroup>`__
4. 🔒 `Irit's Discord post with Core Sprint 2023 slides about C API workgroup <https://discord.com/channels/854719841091715092/897862918115315813/1160952205801689249>`__
5. 🔒 `Petr's Discord post with Core Sprint 2023 slides <https://discord.com/channels/854719841091715092/897862918115315813/1160939833473126411>`__
6. 🔒 `Steve's Discord post with Core Sprint 2023 slides for Things to Learn from HPy <https://discord.com/channels/854719841091715092/897862918115315813/1160942893696024578>`__

> 7. `Victor's slides of Core Sprint 2023 Python C API talk: <https://github.com/vstinner/talks/blob/main/2023-CoreDevSprint-Brno/c-api.pdf>`__

Reviewer: ```suggestion
7. `Victor's slides of Core Sprint 2023 Python C API talk <https://github.com/vstinner/talks/blob/main/2023-CoreDevSprint-Brno/c-api.pdf>`__
```

8. `The Python's stability promise — Cristián Maureira-Fredes, PySide maintainer <https://www.youtube.com/watch?v=iiBJF0kM-P8>`__

> 9. `Report on the issues PySide had 5 years ago when switching to the stable ABI: <https://github.com/pyside/pyside2-setup/blob/5.11/sources/shiboken2/libshiboken/pep384impl_doc.rst>`__

Reviewer: ```suggestion
9. `Report on the issues PySide had 5 years ago when switching to the stable ABI <https://github.com/pyside/pyside2-setup/blob/5.11/sources/shiboken2/libshiboken/pep384impl_doc.rst>`__
```

PEP: 733
Title: An Evaluation of Python's Public C API

> Author: Erlend E. Aasland <erlend@python.org>, Domenico Andreoli <>, Stefan Behnel <stefan_ml@behnel.de>, Carl Friedrich Bolz-Tereick <cfbolz@gmx.de>, Simon Cross <>, Steve Dower <steve.dower@python.org>, Tim Felgentreff <tim.felgentreff@oracle.com>, David Hewitt <>, Shantanu Jain <hauntsaninja at gmail.com>, Wenzel Jakob <>, Irit Katriel <irit@python.org>, Marc-Andre Lemburg <mal@lemburg.com>, Donghee Na <donghee.na@python.org>, Karl Nelson <>, Ronald Oussoren <ronaldoussoren@mac.com>, Matti Picus <>, Antoine Pitrou <solipsis@pitrou.net>, Benedikt Reinartz <>, Guido van Rossum <guido@python.org>, Neil Schemenauer <nas@arctrix.com>, Mark Shannon <mark@hotpy.org>, Stepan Sindelar <stepan.sindelar@oracle.com>, Gregory P. Smith <greg@krypto.org>, Eric Snow <ericsnowcurrently@gmail.com>, Victor Stinner <vstinner@python.org>, Serhiy Storchaka <>, Petr Viktorin <encukou@gmail.com>, Carol Willing <willingc@gmail.com>, William Woodruff <william@yossarian.net>, David Woods <>, Jelle Zijlstra <jelle.zijlstra@gmail.com>

Reviewer: I removed Matti, Benedikt, Antoine, and added missing email addresses:

```suggestion
Author: Erlend E. Aasland <erlend@python.org>, Domenico Andreoli <domenico.andreoli@linux.com>, Stefan Behnel <stefan_ml@behnel.de>, Carl Friedrich Bolz-Tereick <cfbolz@gmx.de>, Simon Cross <hodgestar@gmail.com>, Steve Dower <steve.dower@python.org>, Tim Felgentreff <tim.felgentreff@oracle.com>, David Hewitt <1939362+davidhewitt@users.noreply.github.com>, Shantanu Jain <hauntsaninja at gmail.com>, Wenzel Jakob <wenzel.jakob@epfl.ch>, Irit Katriel <irit@python.org>, Marc-Andre Lemburg <mal@lemburg.com>, Donghee Na <donghee.na@python.org>, Karl Nelson <nelson85@llnl.gov>, Ronald Oussoren <ronaldoussoren@mac.com>, Guido van Rossum <guido@python.org>, Neil Schemenauer <nas@arctrix.com>, Mark Shannon <mark@hotpy.org>, Stepan Sindelar <stepan.sindelar@oracle.com>, Gregory P. Smith <greg@krypto.org>, Eric Snow <ericsnowcurrently@gmail.com>, Victor Stinner <vstinner@python.org>, Serhiy Storchaka <storchaka@gmail.com>, Petr Viktorin <encukou@gmail.com>, Carol Willing <willingc@gmail.com>, William Woodruff <william@yossarian.net>, David Woods <dw-git@d-woods.co.uk>, Jelle Zijlstra <jelle.zijlstra@gmail.com>
```

I got missing email addresses in public Git repositories:

* Domenico Andreoli @cavokz: https://github.com/cavokz/cavokz (commit)
* Simon Cross @hodgestar: https://github.com/hpyproject/hpy/ (commit)
* David Hewitt @davidhewitt: https://github.com/PyO3/pyo3 (commit)
* Wenzel Jakob @wjakob: https://github.com/wjakob/ (email on the profile)
* Karl Nelson @Thrameos: https://github.com/jpype-project/jpype (commit, 2020)
* Benedikt Reinartz @filmor: (asked to not be listed)
* Serhiy Storchaka @serhiy-storchaka: https://github.com/python/cpython (commit)
* David Woods @da-woods: https://github.com/cython/cython/ (commit)

can also be seen as belonging to this category of stakeholders.

Such systems need minimal building blocks for accessing CPython
efficiently. They don't necessarily need an ergonomic API, because
they typically generate code that is not intended to be read
by humans. But they do need it to be comprehensive enough so that
they don't need to access internals, while offering them stability,
and without sacrificing performance.

An alternative is to have a fast API tier with less error checking
and lower stability guarantees. Then the developers and users of
these tools can choose whether to generate code that uses the
faster or the safer and more stable version of the API.

Binding Generators
------------------

Libraries that create bindings between Python and other object models,
paradigms or languages, such as
`pybind11 <https://pybind11.readthedocs.io/en/stable/>`__ for C++11,
`PyO3 <https://github.com/PyO3/pyo3>`__ for Rust,
`PySide <https://pypi.org/project/PySide/>`__ for Qt,
`PyGObject <https://pygobject.readthedocs.io/en/latest/>`__ for GTK,
`Pygolo <https://gitlab.com/pygolo/py>`__ for Go,
`PyJNIus <https://github.com/kivy/pyjnius/>`__ for Java, or

> `SWIG <https://swig.org/>`__ for C/C++.

Reviewer 2: ```suggestion
`SWIG <https://swig.org/>`__ for C/C++.
`Python.NET <https://github.com/pythonnet/pythonnet>`__ for .NET (C#).
```

Python.NET arguably belongs in this category, although its bindings are completely dynamic right now and it's not really a "binding generator".

Reviewer 3: Thanks for the work on Python.NET, by the way!

Reviewer: Added!

> `PyJNIus <https://github.com/kivy/pyjnius/>`__ for Java, or
> `SWIG <https://swig.org/>`__ for C/C++.

Reviewer: ```suggestion
`PyJNIus <https://github.com/kivy/pyjnius/>`__ for Java,
`SWIG <https://swig.org/>`__ for C/C++, or
`Python.NET <https://github.com/pythonnet/pythonnet>`__ for .NET (C#).
```

that its maintenance will follow, as well as the technical and
organizational processes by which this will work.

If the model does include provisions for incremental evolution of the API,
it will include processes for managing the impact of the change on users
[`Issue 60 <https://github.com/capi-workgroup/problems/issues/60>`__],
perhaps through introducing an external backwards compatibility module
[`Issue 62 <https://github.com/capi-workgroup/problems/issues/62>`__],
or a new API tier of "blessed" functions
[`Issue 55 <https://github.com/capi-workgroup/problems/issues/55>`__].

API Specification and Abstraction
---------------------------------

The C API does not have a formal specification, it is described
semi-formally in the documentation and exposed through C header
files. This creates a number of problems.

Bindings for languages other than C/C++ must parse C code
[`Issue 7 <https://github.com/capi-workgroup/problems/issues/7>`__].
Some C language features are hard to handle in this way, because
they produce compiler-dependent output (such as enums) or require
a C preprocessor/compiler rather than just a parser (such as macros)

> [`Issue 35 <https://github.com/capi-workgroup/problems/issues/35>`__].

Reviewer 2: Things that can also be problematic for "less complete" C FFIs like .NET's are
- accessing non-function symbols (`PyNone` etc.)
- functions with variadic arguments

Reviewer: > accessing non-function symbols (PyNone etc.)

Do you think that the PEP doesn't already explain this issue properly? Do you want to propose to rephrase or elaborate a section?

> functions with variadic arguments

see line 507

Reviewer 2: I'll have another look, maybe I missed this.

Reviewer 2: Ah, I see, I'll add non-function symbols to the list as a suggestion :)

PEP: 733
Title: An Evaluation of Python's Public C API

> Author: Erlend E. Aasland <erlend@python.org>, Domenico Andreoli <domenico.andreoli@linux.com>, Stefan Behnel <stefan_ml@behnel.de>, Carl Friedrich Bolz-Tereick <cfbolz@gmx.de>, Simon Cross <hodgestar@gmail.com>, Steve Dower <steve.dower@python.org>, Tim Felgentreff <tim.felgentreff@oracle.com>, David Hewitt <1939362+davidhewitt@users.noreply.github.com>, Shantanu Jain <hauntsaninja at gmail.com>, Wenzel Jakob <wenzel.jakob@epfl.ch>, Irit Katriel <irit@python.org>, Marc-Andre Lemburg <mal@lemburg.com>, Donghee Na <donghee.na@python.org>, Karl Nelson <nelson85@llnl.gov>, Ronald Oussoren <ronaldoussoren@mac.com>, Antoine Pitrou <solipsis@pitrou.net>, Neil Schemenauer <nas@arctrix.com>, Mark Shannon <mark@hotpy.org>, Stepan Sindelar <stepan.sindelar@oracle.com>, Gregory P. Smith <greg@krypto.org>, Eric Snow <ericsnowcurrently@gmail.com>, Victor Stinner <vstinner@python.org>, Serhiy Storchaka <storchaka@gmail.com>, Petr Viktorin <encukou@gmail.com>, Carol Willing <willingc@gmail.com>, William Woodruff <william@yossarian.net>, David Woods <dw-git@d-woods.co.uk>, Jelle Zijlstra <jelle.zijlstra@gmail.com>

Reviewer 4: As others have noticed, this is a _really_ long list of authors for a PEP, much longer than any I can recall before, which seems potentially quite difficult to read and manage, since you'll need to get approval from all of them to post the PEP, and if changes are proposed, how do you ensure everyone agrees on them? However, there isn't any hard limit on author count, so up to you if you want to manage that.

To make that a _bit_ easier, I've at least formatted the header properly. You might also want to sort it lexicographically by given name, surname or equivalent to make it easier to search for and find a specific person; you're welcome to edit my comment if so.

```suggestion
Author: Erlend E. Aasland <erlend@python.org>,
        Domenico Andreoli <domenico.andreoli@linux.com>,
        Stefan Behnel <stefan_ml@behnel.de>,
        Carl Friedrich Bolz-Tereick <cfbolz@gmx.de>,
        Simon Cross <hodgestar@gmail.com>,
        Steve Dower <steve.dower@python.org>,
        Tim Felgentreff <tim.felgentreff@oracle.com>,
        David Hewitt <1939362+davidhewitt@users.noreply.github.com>,
        Shantanu Jain <hauntsaninja at gmail.com>,
        Wenzel Jakob <wenzel.jakob@epfl.ch>,
        Irit Katriel <irit@python.org>,
        Marc-Andre Lemburg <mal@lemburg.com>,
        Donghee Na <donghee.na@python.org>,
        Karl Nelson <nelson85@llnl.gov>,
        Ronald Oussoren <ronaldoussoren@mac.com>,
        Antoine Pitrou <solipsis@pitrou.net>,
        Neil Schemenauer <nas@arctrix.com>,
        Mark Shannon <mark@hotpy.org>,
        Stepan Sindelar <stepan.sindelar@oracle.com>,
        Gregory P. Smith <greg@krypto.org>,
        Eric Snow <ericsnowcurrently@gmail.com>,
        Victor Stinner <vstinner@python.org>,
        Serhiy Storchaka <storchaka@gmail.com>,
        Petr Viktorin <encukou@gmail.com>,
        Carol Willing <willingc@gmail.com>,
        William Woodruff <william@yossarian.net>,
        David Woods <dw-git@d-woods.co.uk>,
        Jelle Zijlstra <jelle.zijlstra@gmail.com>
```

Reviewer 4: Just for fun, using the PEP "API":

```python
import pandas as pd
import requests

peps = requests.get("https://peps.python.org/api/peps.json").json()
df = pd.DataFrame.from_dict(peps, orient="index")
```

I can see the longest existing authors field is 197 characters (as a comm-separated list of just the names):

```pycon
>>> max(df["authors"].str.len())
197
```

This PEP's authors has a length a bit over twice that:

```pycon
>>> len("Erlend E. Aasland, Domenico Andreoli, Stefan Behnel, Carl Friedrich Bolz-Tereick, Simon Cross, Steve Dower, Tim Felgentreff, David Hewitt, Shantanu Jain, Wenzel Jakob, Irit Katriel, Marc-Andre Lemburg, Donghee Na, Karl Nelson, Ronald Oussoren, Antoine Pitrou, Neil Schemenauer, Mark Shannon, Stepan Sindelar, Gregory P. Smith, Eric Snow, Victor Stinner, Serhiy Storchaka, Petr Viktorin, Carol Willing, William Woodruff, David Woods, Jelle Zijlstra")
447
```

Reviewer 5: Your suggestion does make it a lot more readable, thanks. Note that it's already sorted by last name.

There are API functions which have an unsafe (but fast) version as well as
a safe version which performs error checking (for example,
``PyTuple_GET_ITEM`` vs ``PyTuple_GetItem``). It may help to
be able to group them into their own tiers - the "unsafe API" tier and
the "safe API" tier
[`Issue 61 <https://github.com/capi-workgroup/problems/issues/61>`__].

Use of the C Language
---------------------

A number of issues were raised with respect to the way that CPython
uses the C language. First there is the issue of which C dialect
we use, and how we test our compatibility with it
[`Issue 42 <https://github.com/capi-workgroup/problems/issues/42>`__].

Usage of ``const`` in the API is currently sparse, but it is not
clear whether this is something that we should consider changing
[`Issue 38 <https://github.com/capi-workgroup/problems/issues/38>`__].

We currently use the C types ``long`` and ``int``, where fixed-width integers
such as ``int32_t`` and ``int64_t`` would have been better choices
[`Issue 27 <https://github.com/capi-workgroup/problems/issues/27>`__].

We are using C language features which are hard for other languages

> to interact with, such as macros, variadic arguments, enums, and bitfields

Reviewer 2: ```suggestion
to interact with, such as macros, variadic arguments, enums, bitfields,
and non-function symbols
```

Reviewer: I took your suggestion. It's always better to be more explicit :-)

Status: Draft
Type: Informational
Content-Type: text/x-rst
Created: 16-Oct-2023

Abstract
========

This **informational** PEP describes our shared view of the public C API. The
document defines:

* purposes of the C API
* stakeholders and their particular use cases and requirements
* strengths of the C API
* problems of the C API categorized into nine areas of weakness

This document does not propose solutions to any of the identified problems. By
creating a shared list of C API issues, this document will help to guide
continuing discussion about change proposals and to identify evaluation
criteria.

Introduction
============

> Python's C API was not designed for the purposes it currently fulfils.

Reviewer 6: ```suggestion
Python's C API was not designed for the purposes it currently fulfills.
```
The US spelling is usually 'fulfill' rather than 'fulfil' (mostly UK).

Reviewer 5: As Irit is the primary author of the text, I think we should keep the British spellings. I think there's also a "behaviour" further down.

Reviewer: I'm French and learnt UK spelling at school. It's very confusing for me when people report me "typos" in my spelling and suggest the US spelling, whereas I learnt correctly the UK spelling :-) I understood that in Python, most people use the US spelling. In general, I mostly care about consistency. For this document, I'm fine with UK or US, as long as it's consistent in the whole document (as @JelleZijlstra says).

Note: I didn't know that fulfils has different spellings!

Author: My mother tongue is Hebrew, I learned English in the US and now I like in the UK. So expect my English to be a messy mixture of the two.

identified a number of recurring themes. The sections below
mostly correspond to these themes, and each contains a combined
description of the issues raised in that category, along with
links to the individual issues. In addition, we included a section
that aims to identify the different stakeholders of the C API,
and the particular requirements that each of them has.

C API Stakeholders
==================

As mentioned in the introduction, the C API was originally
created as the internal interface between CPython's
interpreter and the Python layer. It was later exposed as
a way for third-party developers to extend and embed Python
programs. Over the years, new types of stakeholders emerged,
with different requirements and areas of focus. This section
describes this complex state of affairs in terms of the
actions that different stakeholders need to perform through
the C API.

Universal Actions for All Stakeholders
--------------------------------------

There are actions which are generic, and required by
all types of API users:

* Define functions and call them
* Create classes and instantiate them
* Create instances of builtin and user-defined types

>   and perform operations on them

Reviewer 6: There is some overlap between these two points, or rather, the second seems to include the first.

* Introspect objects, including types, instances, and functions
* Raise and handle exceptions
* Import modules
* Manage threads
* Manage sub-interpreters

> * Handle and send signals

Author: The C API does have a few [signal related functions](https://docs.python.org/3/c-api/exceptions.html#signal-handling).  Mentioning it here simply means we need to keep them in any redefinition of the API. Do you think this is not the case?

> * Handle and send signals

Reviewer 7: I doubt we can get rid of them completely (though they're only poorly emulated and best avoided on Windows, which is a significant proportion of "all stakeholders").

It's more that I wouldn't define them as "universal actions". An unavoidable aspect of implementing real software on certain platforms, yes, but not inherently a feature of Python. ("Managing threads" is _almost_ the same, but the rest are clearly Python-defined features.)

> * Handle and send signals

Author: I think I understand. How about if we replace "managing threads" and "handle and send signals" by a single entry with something more general, like "access to python's OS interface"?

> * Handle and send signals

Reviewer 7: I'll settle for it 😄

They need to be able to:

* Configure the interpreter (import paths, inittab, ``sys.argv``, etc.).
* Interact with the execution model and program lifetime, including
  clean interpreter shutdown and restart.
* Represent complex data models in a way Python can use without
  having to create deep copies.
* Provide and import frozen modules.
* Run multiple independent interpreters (in particular, when embedded
  in a library that wants to avoid global effects).

Alternative Python Implementations
----------------------------------

Alternative implementations of Python (such as
`PyPy <https://www.pypy.org>`__,
`GraalPy <https://www.graalvm.org/python/>`__,
`IronPython <https://ironpython.net>`__,
`RustPython <https://github.com/RustPython/RustPython>`__,
`MicroPython <https://micropython.org>`__,
and `Jython <https://www.jython.org>`__), may take
very different approaches for the implementation of
different subsystems. They need:

* The API to be abstract and hide implementation details.
* A specification of the API, ideally with a test suite
  that ensures compatibility.
* It would be nice to have an ABI that can be shared

>   across Python implementations.

Author: I'm not sure we are making this distinction between "Universal Actions" which must be compatible with CPython and "additional actions" that may not be.

>   across Python implementations.

Reviewer 7: Let's not call them "Universal" then. (Or alternatively, let's officially decide that CPython is not the reference interpreter... but that's probably still a bit too controversial...)

We _ought_ to clearly distinguish between the intended goals ("universal Python API") and what we're evaluating here ("CPython's API", an imperfect superset of the universal Python API). There's just no way to define them as the same thing unless we immediately start ignoring that definition.

A really easy example is that we can't expect the initialization APIs to be identical between all implementations. So does that mean that initialization is not "universal" (my preference)? Or that we say initialization _is_ universal and everyone besides CPython is just incompatible with the universal Python API?

But my main point here is that every other group listed in this section has a list of jobs to be done, while this group has a list of aspects of the API that they hope to be true. It's inconsistent, and I think it would be better to list this group's jobs to be done (specifically, "implement the Universal Actions") before getting into their hopes and dreams for the ABI.

>   across Python implementations.

Author: "Universal" here was meant as "relevant to all stakeholders" (rather than a subset of the stakeholders). 

Regarding the requirements we list: they are not "what this stakeholder need to do" but rather "what this stakeholders needs from the C API". So "implement the Universal Actions" is not appropriate (or at least not specific enough). 

I see your point that the requirements of alternative implementations are of a different nature to those of people who use the API to implement something else. They are indeed a different type of stakeholder. An alternative implementation doesn't need to use the C API. It needs to know what it is so that it can implement it.  It doesn't need anything functional from the C API, it needs a good spec of what it does.

>   across Python implementations.

Reviewer 8: > we can't expect the initialization APIs to be identical between all implementations

I assume you mean initialization of embedded (C)Python? I do not see any inherent reason why it cannot be universal. Java native interface (JNI) provides embedding API that is implemented by several different Java virtual machines.

(The initialization of extensions can be universal, which is demonstrated by HPy).

>   across Python implementations.

Author: Maybe universal was the wrong word. Would 'generic' or 'basic' be better?

>   across Python implementations.

Reviewer 9: Maybe "Common" or "Fundamental" @iritkatriel. 'Generic' works too.

>   across Python implementations.

Author: "Common" is good. Thanks.

Alternative APIs
----------------

There are several projects that implement alternatives to the
C API, which offer extension users advantanges over programming
directly with the C API. These APIs are implemented with the
C API, and in some cases by using CPython internals.
Some examples are
`Cython <https://cython.org>`__,
`HPy <https://hpyproject.org>`__ and
`pythoncapi-compat <https://pythoncapi-compat.readthedocs.io/en/latest/>`__.
CPython's DSL for parsing function arguments, the
`Argument Clinic <https://docs.python.org/3/howto/clinic.html>`__,
can also be seen as belonging to this category of stakeholders.

Such systems need minimal building blocks for accessing CPython
efficiently. They don't necessarily need an ergonomic API, because
they typically generate code that is not intended to be read
by humans. But they do need it to be comprehensive enough so that
they don't need to access internals, while offering them stability,
and without sacrificing performance.

An alternative is to have a fast API tier with less error checking
and lower stability guarantees. Then the developers and users of
these tools can choose whether to generate code that uses the

> faster or the safer and more stable version of the API.

Reviewer 7: "An alternative" to what?

Do we mean: "These developers might be best satisfied with a choice between a faster but less stable API tier"? Which seems adequately covered by the previous paragraph to me.

Perhaps we should be continuing to list desirable actions for these tools? Such as:
* efficiently convert between primitive/native values and Python objects
* track ownership of object references
* detect availability of stable and performant APIs at compile-time and/or runtime

Binding Generators
------------------

Libraries that create bindings between Python and other object models,
paradigms or languages, such as
`pybind11 <https://pybind11.readthedocs.io/en/stable/>`__ for C++11,
`PyO3 <https://github.com/PyO3/pyo3>`__ for Rust,
`PySide <https://pypi.org/project/PySide/>`__ for Qt,
`PyGObject <https://pygobject.readthedocs.io/en/latest/>`__ for GTK,
`Pygolo <https://gitlab.com/pygolo/py>`__ for Go,
`PyJNIus <https://github.com/kivy/pyjnius/>`__ for Java,
`SWIG <https://swig.org/>`__ for C/C++, or
`Python.NET <https://github.com/pythonnet/pythonnet>`__ for .NET (C#).

Some of the alternative APIs mentioned in the previous section create
alternative object models and therefore they can also be seen as
belonging to this category.

> They need to:

Reviewer 7: Having now read this section, I think `Cython` belongs here and not in the previous one.

Higher-level abstractions over the C API (the previous section) are probably a less important consideration, considering they're largely designed to cover over the flaws/limitations of the C API that we want to fix. We should be aiming to make them mostly redundant, rather than trying to actively support them (e.g. if we had designed it perfectly in the first place, HPy and pythoncapi-compat would never have existed).

Cython is different enough from our C API that we aren't going to make it redundant, so I think it belongs here.

Reviewer 10: IMO, Cython belongs in both. Many of these do, actually.
Since these are non-exhaustive lists, it's OK to only put it in one place.

Reviewer 7: > Since these are non-exhaustive lists, it's OK to only put it in one place.

We're not concerned about the exhaustiveness of the lists, but that Cython is currently in the _wrong_ list. Non-exhaustive doesn't mean you can incorrectly categorise the items that you do include.

perhaps through introducing an external backwards compatibility module
[`Issue 62 <https://github.com/capi-workgroup/problems/issues/62>`__],
or a new API tier of "blessed" functions
[`Issue 55 <https://github.com/capi-workgroup/problems/issues/55>`__].

API Specification and Abstraction
---------------------------------

The C API does not have a formal specification, it is described
semi-formally in the documentation and exposed through C header
files. This creates a number of problems.

Bindings for languages other than C/C++ must parse C code
[`Issue 7 <https://github.com/capi-workgroup/problems/issues/7>`__].
Some C language features are hard to handle in this way, because
they produce compiler-dependent output (such as enums) or require
a C preprocessor/compiler rather than just a parser (such as macros)
[`Issue 35 <https://github.com/capi-workgroup/problems/issues/35>`__].

Furthermore, C header files tend to expose more than what is intended
to be part of the public API
[`Issue 34 <https://github.com/capi-workgroup/problems/issues/34>`__].
In particular, implementation details such as the fields of C structs

> can be exposed

Reviewer 7: Strictly speaking, the precise memory layout of our internal data structures is being exposed. (Sometimes we expose C structs on purpose.) I suspect the distinction doesn't matter that much though.

Author: reworded.

Several aspects of the type definition and creation process are not
well defined, such as which stage of the process is responsible for
initializing and clearing different fields of the type object
[`Issue 49 <https://github.com/capi-workgroup/problems/issues/49>`__].

Error Handling
--------------

Error handling in the C API is based on the error indicator which is stored
on the thread state (in global scope). The design intention was that each
API function returns a value indicating whether an error has occurred (by
convention, ``-1`` or ``NULL``). When the program knows that an error
occurred, it can fetch the exception object which is stored in the
error indicator. We identified a number of problems which are related
to error handling, pointing at APIs which are too easy to use incorrectly.

There are functions that do not report all errors that occur while they
execute. For example, ``PyDict_GetItem`` clears any errors that occur
when it calls the key's hash function, or while performing a lookup
in the dictionary
[`Issue 51 <https://github.com/capi-workgroup/problems/issues/51>`__].

Python code never executes with an in-flight exception (by definition),

> and by the same token C API functions should never be called with the error
> indicator set. This is currently not checked in most C API functions, and

Reviewer 7: I think this is presuming a design that we haven't really discussed. While it's bad for an API to lose existing exception information, there's no reason to say "should never" here at this stage (and obviously I don't think we should say "never" at any stage, but we can discuss it).

How about:
```suggestion
and typically code using native functions should also be interrupted by
an error being raised. This is not checked in most C API functions, and
```

However, this was only clarified recently, in :pep:`689`. It is
not clear what the change policy should be with respect to such
API elements that predate PEP 689
[`Issue 58 <https://github.com/capi-workgroup/problems/issues/58>`__].

There are API functions which have an unsafe (but fast) version as well as
a safe version which performs error checking (for example,
``PyTuple_GET_ITEM`` vs ``PyTuple_GetItem``). It may help to
be able to group them into their own tiers - the "unsafe API" tier and
the "safe API" tier
[`Issue 61 <https://github.com/capi-workgroup/problems/issues/61>`__].

Use of the C Language
---------------------

A number of issues were raised with respect to the way that CPython
uses the C language. First there is the issue of which C dialect
we use, and how we test our compatibility with it
[`Issue 42 <https://github.com/capi-workgroup/problems/issues/42>`__].

Usage of ``const`` in the API is currently sparse, but it is not
clear whether this is something that we should consider changing
[`Issue 38 <https://github.com/capi-workgroup/problems/issues/38>`__].

We currently use the C types ``long`` and ``int``, where fixed-width integers

> such as ``int32_t`` and ``int64_t`` would have been better choices

Reviewer 7: ```suggestion
such as ``int32_t`` and ``int64_t`` may now be better choices
```

These were only added in C99, so weren't viable options until recently. No reason we should suggest the original design was incorrect, it was perfect for its time.

[`Issue 25 <https://github.com/capi-workgroup/problems/issues/25>`__].

The macros ``Py_CLEAR`` and ``Py_SETREF`` access their arg more than
once, so if the arg is an expression with side effects, they are
duplicated
[`Issue 3 <https://github.com/capi-workgroup/problems/issues/3>`__].

The meaning of ``Py_SIZE`` depends on the type and is not always
reliable
[`Issue 10 <https://github.com/capi-workgroup/problems/issues/10>`__].

Some API function do not have the same behaviour as their Python
equivalents.  The behaviour of ``PyIter_Next`` is different from
``tp_iternext``.
[`Issue 29 <https://github.com/capi-workgroup/problems/issues/29>`__].
The behaviour of ``PySet_Contains`` is different from ``set.__contains__``
[`Issue 6 <https://github.com/capi-workgroup/problems/issues/6>`__].

The fact that ``PyArg_ParseTupleAndKeywords`` takes a non-const
``char*`` array as argument makes it more difficult to use
[`Issue 28 <https://github.com/capi-workgroup/problems/issues/28>`__].

``Python.h`` does not expose the whole API. Some headers (like ``marshal.h``)
are not included from ``Python.h``.
[`Issue 43 <https://github.com/capi-workgroup/problems/issues/43>`__].

**Naming**

> ``PyLong`` and ``PyUnicode`` use names which don't match the Python

Reviewer 7: Is it worth specifying that they _did_ match the type names back when they were added? In the spirit of not unduly criticising the original API designers (many of whom are still around and are likely to read this document).

Reviewer 6: IE, change "don't" to "no longer", which implies "once did".

types they represent (``int``/``str``). This can be fixed in a new API
[`Issue 14 <https://github.com/capi-workgroup/problems/issues/14>`__].

There are identifiers in the API which are lacking a ``Py``/``_Py``
prefix
[`Issue 46 <https://github.com/capi-workgroup/problems/issues/46>`__].

Missing Functionality
---------------------

This section consists of a list of feature requests, i.e., functionality
that was identified as missing in the current C API.

Debug Mode
~~~~~~~~~~

A debug mode that can be activated without recompilation and which
activates various checks that can help detect various types of errors
[`Issue 36 <https://github.com/capi-workgroup/problems/issues/36>`__].

Introspection
~~~~~~~~~~~~~

There aren't currently reliable introspection capabilities for objects
defined in C in the same way as there are for Python objects
[`Issue 32 <https://github.com/capi-workgroup/problems/issues/32>`__].

Efficient type checking for heap types, similar to what ``Py*_Check``

> can do for a static type

Reviewer 7: Used to be able to do - they're equally slow now, right? Only efficient when we know that subclassing isn't possible.

Reviewer 11: Last I looked, the C-API type checks for builtin types were actually efficiently using type flags.

Reviewer 7: "Static" is not the same as "builtin". The ones with flags are more special than static types.

Maybe built-in was intended here? I didn't want to presume that. But it's very unlikely we'll be able to give all types a flag that is inherited by subclasses, so if that's the intention, I'd vote to rule out the suggestion now.

Reviewer 7: Yeah, they still can, but arbitrary static types could not be checked using a bitmask - only those that inherited from a builtin type.

Reviewer 3: Ah, right. Well, in any case, efficient type checking is important for several applications, to the point that I committed [this horror](https://github.com/apache/arrow/blob/d68f8e21643e8aec9ed253094edfd15c0a08f1c1/python/pyarrow/src/arrow/python/helpers.cc#L277-L284) at some point.

Reviewer 7: I don't disagree, but the cases here are dictated by Python's semantics and not by the C API:
* you have a `type` object and an instance of that type (fast)
* you have a `type` object and an instance of a subclass of that type (slow)
* you have a `type` object and an instance of a type that implements `__instancecheck__` or `__subclasscheck__` (slower)
* you don't have a `type` object (umm... not sure how to generalise type checking when you don't have a type)

Boolean flags are an optimisation for certain built-in types only, we can't use them as the expectation for all types. The general case (i.e. for anything not built into the runtime) is a C version of `isinstance(o, T)` or `type(o) is T`.

Reviewer 3: Well, whether the general case can be made fast is an open question, though there is no reason that it couldn't.

Reviewer 10: There are also two kinds of type checking: `PyObject_IsSubclass` (which allows `__*check__` hooks) and `PyType_IsSubtype` (which should guarantee compatible memory layout, i.e. whether it's safe to cast).
In the C-API you usually care about the latter. We could allow checking compatible memory layout without an actual `type` object to compare to (see [#17](https://github.com/capi-workgroup/problems/issues/17)), and you could say that's just a workaround for an even thornier problem -- but IMO, that level of detail is out of scope for this PEP.

Author: Thanks all. I'll just remove the contentious part of this sentence, we can do without it.

[`Issue 17 <https://github.com/capi-workgroup/problems/issues/17>`__].

Improved Interaction with Other Languages
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Interfacing with other GC based languages, and integrating their
GC with Python's GC
[`Issue 19 <https://github.com/capi-workgroup/problems/issues/19>`__].

Inject foreign stack frames to the traceback
[`Issue 18 <https://github.com/capi-workgroup/problems/issues/18>`__].

Concrete strings that can be used in other languages
[`Issue 16 <https://github.com/capi-workgroup/problems/issues/16>`__].

References
==========

1. `Python/C API Reference Manual <https://docs.python.org/3/c-api/index.html>`__
2. `2023 Language Summit Blog Post: Three Talks on the C API <https://pyfound.blogspot.com/2023/05/the-python-language-summit-2023-three.html>`__
3. `capi-workgroup on GitHub <https://github.com/capi-workgroup>`__

> 4. 🔒 `Irit's Discord post with Core Sprint 2023 slides about C API workgroup <https://discord.com/channels/854719841091715092/897862918115315813/1160952205801689249>`__

Reviewer: ```suggestion
4. `Irit's Core Sprint 2023 slides about C API workgroup <https://github.com/iritkatriel/talks/blob/main/2023_Sprint_Brno_C_API.pdf>`__
```

> 5. 🔒 `Petr's Discord post with Core Sprint 2023 slides <https://discord.com/channels/854719841091715092/897862918115315813/1160939833473126411>`__

Reviewer 10: ```suggestion
5. `Petr's Core Sprint 2023 slides <https://drive.google.com/file/d/148NLRPXGZGI1SXfKLMFvQc_iv67hPJQS/view?usp=sharing>`__
```

> 6. 🔒 `Steve's Discord post with Core Sprint 2023 slides for Things to Learn from HPy <https://discord.com/channels/854719841091715092/897862918115315813/1160942893696024578>`__

Reviewer 7: These are not public links - will only be accessible by those with permission. Is that an okay reference to include in a public doc?

Reviewer 12: Irit's and Petr's slides are public and the links have been updated:

https://pep-previews--3491.org.readthedocs.build/pep-0733/#references

And the Hpy slides will be public soon.

PEP: 733
Title: An Evaluation of Python's Public C API

> Author: Erlend E. Aasland <erlend@python.org>, Domenico Andreoli <>, Stefan Behnel <stefan_ml@behnel.de>, Carl Friedrich Bolz-Tereick <cfbolz@gmx.de>, Simon Cross <>, Steve Dower <steve.dower@python.org>, Tim Felgentreff <tim.felgentreff@oracle.com>, David Hewitt <>, Shantanu Jain <hauntsaninja at gmail.com>, Wenzel Jakob <>, Irit Katriel <irit@python.org>, Marc-Andre Lemburg <mal@lemburg.com>, Dong-hee Na <>, Karl Nelson <>, Ronald Oussoren <ronaldoussoren@mac.com>, Matti Picus <>, Antoine Pitrou <solipsis@pitrou.net>, Benedikt Reinartz <>, Guido van Rossum <guido@python.org>, Neil Schemenauer <nas@arctrix.com>, Mark Shannon <mark@hotpy.org>, Stepan Sindelar <stepan.sindelar@oracle.com>, Gregory P. Smith <greg@krypto.org>, Eric Snow <ericsnowcurrently@gmail.com>, Victor Stinner <vstinner@python.org>, Serhiy Storchaka <>, Petr Viktorin <encukou@gmail.com>, Carol Willing <willingc@gmail.com>, William Woodruff <william@yossarian.net>, David Woods <>, Jelle Zijlstra <jelle.zijlstra@gmail.com>

Reviewer: I'm not sure that all people who contributed to https://github.com/capi-workgroup/problems/#problems-in-the-python-c-api project expected to get their name and email address listed here. The project README says nothing about a PEP.

Listing all people is great. I just suggest to move the list outside the Authors field. And maybe without email addresses, but maybe with a link to their GitHub profile, since GitHub was used? I don't know.

Author: I only put emails that are already on other PEPs (so already public) or were specifically given for this purpose.

> Author: Erlend E. Aasland <erlend@python.org>, Domenico Andreoli <>, Stefan Behnel <stefan_ml@behnel.de>, Carl Friedrich Bolz-Tereick <cfbolz@gmx.de>, Simon Cross <>, Steve Dower <steve.dower@python.org>, Tim Felgentreff <tim.felgentreff@oracle.com>, David Hewitt <>, Shantanu Jain <hauntsaninja at gmail.com>, Wenzel Jakob <>, Irit Katriel <irit@python.org>, Marc-Andre Lemburg <mal@lemburg.com>, Dong-hee Na <>, Karl Nelson <>, Ronald Oussoren <ronaldoussoren@mac.com>, Matti Picus <>, Antoine Pitrou <solipsis@pitrou.net>, Benedikt Reinartz <>, Guido van Rossum <guido@python.org>, Neil Schemenauer <nas@arctrix.com>, Mark Shannon <mark@hotpy.org>, Stepan Sindelar <stepan.sindelar@oracle.com>, Gregory P. Smith <greg@krypto.org>, Eric Snow <ericsnowcurrently@gmail.com>, Victor Stinner <vstinner@python.org>, Serhiy Storchaka <>, Petr Viktorin <encukou@gmail.com>, Carol Willing <willingc@gmail.com>, William Woodruff <william@yossarian.net>, David Woods <>, Jelle Zijlstra <jelle.zijlstra@gmail.com>

Reviewer 13: nit, Legal name was updated.

Donghee Na <donghee.na@python.org>

> Author: Erlend E. Aasland <erlend@python.org>, Domenico Andreoli <>, Stefan Behnel <stefan_ml@behnel.de>, Carl Friedrich Bolz-Tereick <cfbolz@gmx.de>, Simon Cross <>, Steve Dower <steve.dower@python.org>, Tim Felgentreff <tim.felgentreff@oracle.com>, David Hewitt <>, Shantanu Jain <hauntsaninja at gmail.com>, Wenzel Jakob <>, Irit Katriel <irit@python.org>, Marc-Andre Lemburg <mal@lemburg.com>, Dong-hee Na <>, Karl Nelson <>, Ronald Oussoren <ronaldoussoren@mac.com>, Matti Picus <>, Antoine Pitrou <solipsis@pitrou.net>, Benedikt Reinartz <>, Guido van Rossum <guido@python.org>, Neil Schemenauer <nas@arctrix.com>, Mark Shannon <mark@hotpy.org>, Stepan Sindelar <stepan.sindelar@oracle.com>, Gregory P. Smith <greg@krypto.org>, Eric Snow <ericsnowcurrently@gmail.com>, Victor Stinner <vstinner@python.org>, Serhiy Storchaka <>, Petr Viktorin <encukou@gmail.com>, Carol Willing <willingc@gmail.com>, William Woodruff <william@yossarian.net>, David Woods <>, Jelle Zijlstra <jelle.zijlstra@gmail.com>

Author: ```suggestion
Author: Erlend E. Aasland <erlend@python.org>, Domenico Andreoli <>, Stefan Behnel <stefan_ml@behnel.de>, Carl Friedrich Bolz-Tereick <cfbolz@gmx.de>, Simon Cross <>, Steve Dower <steve.dower@python.org>, Tim Felgentreff <tim.felgentreff@oracle.com>, David Hewitt <>, Shantanu Jain <hauntsaninja at gmail.com>, Wenzel Jakob <>, Irit Katriel <irit@python.org>, Marc-Andre Lemburg <mal@lemburg.com>, Donghee Na <donghee.na@python.org>, Karl Nelson <>, Ronald Oussoren <ronaldoussoren@mac.com>, Matti Picus <>, Antoine Pitrou <solipsis@pitrou.net>, Benedikt Reinartz <>, Guido van Rossum <guido@python.org>, Neil Schemenauer <nas@arctrix.com>, Mark Shannon <mark@hotpy.org>, Stepan Sindelar <stepan.sindelar@oracle.com>, Gregory P. Smith <greg@krypto.org>, Eric Snow <ericsnowcurrently@gmail.com>, Victor Stinner <vstinner@python.org>, Serhiy Storchaka <>, Petr Viktorin <encukou@gmail.com>, Carol Willing <willingc@gmail.com>, William Woodruff <william@yossarian.net>, David Woods <>, Jelle Zijlstra <jelle.zijlstra@gmail.com>
```

> Author: Erlend E. Aasland <erlend@python.org>, Domenico Andreoli <>, Stefan Behnel <stefan_ml@behnel.de>, Carl Friedrich Bolz-Tereick <cfbolz@gmx.de>, Simon Cross <>, Steve Dower <steve.dower@python.org>, Tim Felgentreff <tim.felgentreff@oracle.com>, David Hewitt <>, Shantanu Jain <hauntsaninja at gmail.com>, Wenzel Jakob <>, Irit Katriel <irit@python.org>, Marc-Andre Lemburg <mal@lemburg.com>, Donghee Na <donghee.na@python.org>, Karl Nelson <>, Ronald Oussoren <ronaldoussoren@mac.com>, Matti Picus <>, Antoine Pitrou <solipsis@pitrou.net>, Benedikt Reinartz <>, Guido van Rossum <guido@python.org>, Neil Schemenauer <nas@arctrix.com>, Mark Shannon <mark@hotpy.org>, Stepan Sindelar <stepan.sindelar@oracle.com>, Gregory P. Smith <greg@krypto.org>, Eric Snow <ericsnowcurrently@gmail.com>, Victor Stinner <vstinner@python.org>, Serhiy Storchaka <>, Petr Viktorin <encukou@gmail.com>, Carol Willing <willingc@gmail.com>, William Woodruff <william@yossarian.net>, David Woods <>, Jelle Zijlstra <jelle.zijlstra@gmail.com>

Reviewer 14: Please remove me as an author. I don't feel I contributed anything meaningful.

```suggestion
Author: Erlend E. Aasland <erlend@python.org>, Domenico Andreoli <>, Stefan Behnel <stefan_ml@behnel.de>, Carl Friedrich Bolz-Tereick <cfbolz@gmx.de>, Simon Cross <>, Steve Dower <steve.dower@python.org>, Tim Felgentreff <tim.felgentreff@oracle.com>, David Hewitt <>, Shantanu Jain <hauntsaninja at gmail.com>, Wenzel Jakob <>, Irit Katriel <irit@python.org>, Marc-Andre Lemburg <mal@lemburg.com>, Donghee Na <donghee.na@python.org>, Karl Nelson <>, Ronald Oussoren <ronaldoussoren@mac.com>, Antoine Pitrou <solipsis@pitrou.net>, Benedikt Reinartz <>, Guido van Rossum <guido@python.org>, Neil Schemenauer <nas@arctrix.com>, Mark Shannon <mark@hotpy.org>, Stepan Sindelar <stepan.sindelar@oracle.com>, Gregory P. Smith <greg@krypto.org>, Eric Snow <ericsnowcurrently@gmail.com>, Victor Stinner <vstinner@python.org>, Serhiy Storchaka <>, Petr Viktorin <encukou@gmail.com>, Carol Willing <willingc@gmail.com>, William Woodruff <william@yossarian.net>, David Woods <>, Jelle Zijlstra <jelle.zijlstra@gmail.com>
```

* Represent complex data models in a way Python can use without
  having to create deep copies.
* Provide and import frozen modules.
* Run multiple independent interpreters (in particular, when embedded
  in a library that wants to avoid global effects).

Alternative Python Implementations
----------------------------------

Alternative implementations of Python (such as
`PyPy <https://www.pypy.org>`__,
`GraalPy <https://www.graalvm.org/python/>`__,
`IronPython <https://ironpython.net>`__,
`RustPython <https://github.com/RustPython/RustPython>`__,
`MicroPython <https://micropython.org>`__,
and `Jython <https://www.jython.org>`__), may take
very different approaches for the implementation of
different subsystems. They need:

* The API to be abstract and hide implementation details.
* A specification of the API, ideally with a test suite
  that ensures compatibility.
* It would be nice to have an ABI that can be shared
  across Python implementations.

Alternative APIs
----------------

There are several projects that implement alternatives to the
C API, which offer extension users advantanges over programming
directly with the C API. These APIs are implemented with the

> C API, and in some cases by using cpython internals.

Reviewer 5: ```suggestion
C API, and in some cases by using CPython internals.
```

Some examples are
`Cython <https://cython.org>`__,
`HPy <https://hpyproject.org>`__ and
`pythoncapi-compat <https://pythoncapi-compat.readthedocs.io/en/latest/>`__.
CPython's DSL for parsing function arguments, the
`Argument Clinic <https://docs.python.org/3/howto/clinic.html>`__,
can also be seen as belonging to this category of stakeholders.

Such systems need minimal building blocks for accessing CPython
efficiently. They don't necessarily need an ergonomic API, because
they typically generate code that is not intended to be read
by humans. But they do need it to be comprehensive enough so that
they don't need to access internals, while offering them stability,
and without sacrificing performance.

An alternative is to have a fast API tier with less error checking
and lower stability guarantees. Then the developers and users of
these tools can choose whether to generate code that uses the
faster or the safer and more stable version of the API.

Binding Generators
------------------

Libraries that create bindings between Python and other object models,

> paradigms or languages, such as

Reviewer 15: Arguably Cython is a binding generator too (although I don't know if it's worth changing the text). Certainly point 1 below

> Create custom objects (e.g. function/module objects
  and traceback entries) that match the behavior of equivalent
  Python code as closely as possible.

is a big priority for Cython

make sure that they are preserved.

As mentioned in the introduction, the C API enabled the
development and growth of the Python ecosystem over the last
three decades, while evolving to support use cases that it was
not originally designed for. This track record in itself is
indication of how effective and valuable it has been.

A number of specific strengths were mentioned in the
capi-workgroup discussions. Heap types were identified
as much safer and easier to use than static types
[`Issue 4 <https://github.com/capi-workgroup/problems/issues/4#issuecomment-1542324451>`__].

API functions that take a C string literal for lookups based
on a Python string are very convenient
[`Issue 30 <https://github.com/capi-workgroup/problems/issues/30#issuecomment-1550098113>`__].

The Limited API and stable ABI hide implementation details and
make it easier to evolve Python
[`Issue 30 <https://github.com/capi-workgroup/problems/issues/30#issuecomment-1560083258>`__].

C API problems
==============

The remainder of this document summarizes and categorizes the problems that were reported on
the `capi-workgroup <https://github.com/capi-workgroup/problems/issues/>`__ repository.

> The issues are grouped into nine categories:
> 
> - API Evolution and Maintenance
> - API Specification and Abstraction
> - Object Reference Management
> - Type Definition and Object Creation
> - Error Handling
> - API Tiers and Stability Guarantees
> - Use of the C Language
> - Implementation Flaws
> - Missing Functionality

Reviewer 5: ```suggestion
The issues are grouped into several categories.
```
The list feels redundant with the headers (Sphinx will generate a TOC), and "nine" is prone to get out of date.

Object Reference Management
---------------------------

There are C API functions that return borrowed references, and
functions that steal references to arguments, but there isn't a
naming convention that makes this obvious, so this is error prone
[`Issue 8 <https://github.com/capi-workgroup/problems/issues/8>`__
and `Issue 52 <https://github.com/capi-workgroup/problems/issues/52>`__].
The terminology used to describe these situations in the documentation
can also be improved
[`Issue 11 <https://github.com/capi-workgroup/problems/issues/11>`__].

A more radical change is necessary in the case of functions that
return borrowed references (such as ``PyList_GetItem``)
[`Issue 5 <https://github.com/capi-workgroup/problems/issues/5>`__ and
`Issue 21 <https://github.com/capi-workgroup/problems/issues/21>`__]
or pointers to parts of the internal structure of an object
(such as ``PyBytes_AsString``)
[`Issue 57 <https://github.com/capi-workgroup/problems/issues/57>`__].
In both cases, the reference/pointer is valid for as long as the
owning object is alive, but this time is hard to reason about. Such
functions should not exist in the API without a mechanism that can
make them safe.

> For containers, the API is currently missing bulk operations on the

Reviewer 3: Nit: I would call them "batched" operations and also suggest that the batch size and starting point is configurable by the function call.

Author: Can you suggest an edit?

Reviewer 5: I think the current wording is fine. This paragraph describes a problem; it doesn't have to outline the exact API for the solution.

Reviewer 3: Fair enough :-)

private, essentially an API tier with no stability guarantees.
However, this was only clarified recently, in :pep:`689`. It is
not clear what the change policy should be with respect to such
API elements that predate PEP 689
[`Issue 58 <https://github.com/capi-workgroup/problems/issues/58>`__].

There are API functions which have an unsafe (but fast) version as well as
a safe version which performs error checking (for example,
``PyTuple_GET_ITEM`` vs ``PyTuple_GetItem``). It may help to
be able to group them into their own tiers - the "unsafe API" tier and
the "safe API" tier
[`Issue 61 <https://github.com/capi-workgroup/problems/issues/61>`__].

Use of the C Language
---------------------

A number of issues were raised with respect to the way that CPython
uses the C language. First there is the issue of which C dialect
we use, and how we test our compatibility with it
[`Issue 42 <https://github.com/capi-workgroup/problems/issues/42>`__].

Usage of ``const`` in the API is currently sparse, but it is not
clear whether this is something that we should consider changing
[`Issue 38 <https://github.com/capi-workgroup/problems/issues/38>`__].

> We currently use the C types ``long`` and ``int``, where ``stdint``
> and ``int32_t`` would have been better choices

Reviewer 3: I am not sure what `stdint` is supposed to mean here, perhaps you mean fixed-width integers in general?
```suggestion
We currently use the C types ``long`` and ``int``, where fixed-width integers
such as ``int32_t`` and ``int64_t`` would have been better choices
```

[`Issue 27 <https://github.com/capi-workgroup/problems/issues/27>`__].

We are using C language features which are hard for other languages

> to interact with

Reviewer 5: ```suggestion
to interact with, such as macros, variadic arguments, enums, and bitfields
```
Make this more concrete

[`Issue 35 <https://github.com/capi-workgroup/problems/issues/35>`__].

There are API functions that take a ``PyObject*`` arg which must be
of a more specific type (such as ``PyTuple_Size``, which fails if
its arg is not a ``PyTupleObject*``). It is an open question whether this
is a good pattern to have, or whether the API should expect the
more specific type
[`Issue 31 <https://github.com/capi-workgroup/problems/issues/31>`__].

There are functions in the API that take concrete types, such as
``PyDict_GetItemString`` which performs a dictionary lookup for a key

> specified as a c string rather than ``PyObject*``. At the same time,

Reviewer 5: ```suggestion
specified as a C string rather than ``PyObject*``. At the same time,
```

--------------------

Below is a list of localized implementation flaws. Most of these can
probably be fixed incrementally, if we choose to do so. They should,
in any case, be avoided in any new API design.

There are functions that don't follow the convention of
returning ``0`` for success and ``-1`` for failure. For
example, ``PyArg_ParseTuple`` returns 0 for success and
non-zero for failure
[`Issue 25 <https://github.com/capi-workgroup/problems/issues/25>`__].

The macros ``Py_CLEAR`` and ``Py_SETREF`` access their arg more than
once, so if the arg is an expression with side effects, they are
duplicated
[`Issue 3 <https://github.com/capi-workgroup/problems/issues/3>`__].

The meaning of ``Py_SIZE`` depends on the type and is not always
reliable
[`Issue 10 <https://github.com/capi-workgroup/problems/issues/10>`__].

**Naming**

``PyLong`` and ``PyUnicode`` use names which don't match the Python
types they represent (``int``/``str``). This can be fixed in a new API
[`Issue 14 <https://github.com/capi-workgroup/problems/issues/14>`__].

There are identifiers in the API which are lacking a ``Py``/``_Py``
prefix
[`Issue 46 <https://github.com/capi-workgroup/problems/issues/46>`__].

> Some API function do not have the same behaviour as their Python

Reviewer 5: This and the next two paragraphs don't fit under the "Naming" header above. I think they should be moved above that header.

Alternative Python Implementations
----------------------------------

Alternative implementations of Python (such as
`PyPy <https://www.pypy.org>`__,
`GraalPy <https://www.graalvm.org/python/>`__,
`IronPython <https://ironpython.net>`__,
`RustPython <https://github.com/RustPython/RustPython>`__,
`MicroPython <https://micropython.org>`__,
and `Jython <https://www.jython.org>`__), may take
very different approaches for the implementation of
different subsystems. They need:

* The API to be abstract and hide implementation details.
* A specification of the API, ideally with a test suite
  that ensures compatibility.
* It would be nice to have an ABI that can be shared
  across Python implementations.

Alternative APIs
----------------

There are several projects that implement alternatives to the
C API, which offer extension users advantanges over programming
directly with the C API. These APIs are implemented with the
C API, and in some cases by using CPython internals.
Examples are
`HPy <https://hpyproject.org>`__ and
`pythoncapi-compat <https://pythoncapi-compat.readthedocs.io/en/latest/>`__.
CPython's DSL for parsing function arguments, the

> `Argument Clinic <https://docs.python.org/3/howto/clinic.html>`__,

Reviewer 16: Nitpick: the link here is a human redirect to <https://devguide.python.org/development-tools/clinic/>, so it might be worth just direct linking there instead 🙂

Reviewer: Thanks. I updated the link.

> `Argument Clinic <https://docs.python.org/3/howto/clinic.html>`__,

Reviewer: ```suggestion
`Argument Clinic <https://devguide.python.org/development-tools/clinic/>`__,
```

* Introspect objects, including types, instances, and functions
* Raise and handle exceptions
* Import modules
* Manage threads
* Manage sub-interpreters
* Handle and send signals

The following sections look at the unique requirements of various stakeholders.

Extension Writers
-----------------

Extension writers are the traditional users of the C API. Their requirements
are the universal actions listed above.

Authors of Embedded Python Applications
---------------------------------------

Applications with an embedded Python interpreter. Examples are
`Blender <https://docs.blender.org/api/current/info_overview.html>`__ and
`OBS <https://obsproject.com/wiki/Getting-Started-With-OBS-Scripting>`__.

They need to be able to:

* Configure the interpreter (import paths, inittab, ``sys.argv``, etc.).
* Interact with the execution model and program lifetime, including
  clean interpreter shutdown and restart.
* Represent complex data models in a way Python can use without
  having to create deep copies.
* Provide and import frozen modules.
* Run multiple independent interpreters (in particular, when embedded
  in a library that wants to avoid global effects).

> Alternative Python Implementations
> ----------------------------------
> 
> Alternative implementations of Python (such as
> `PyPy <https://www.pypy.org>`__,
> `GraalPy <https://www.graalvm.org/python/>`__,
> `IronPython <https://ironpython.net>`__,
> `RustPython <https://github.com/RustPython/RustPython>`__,
> `MicroPython <https://micropython.org>`__,
> and `Jython <https://www.jython.org>`__), may take
> very different approaches for the implementation of
> different subsystems. They need:

Reviewer: ```suggestion
Python Implementations
----------------------

Python implementations such as
`CPython <https://www.python.org>`__,
`PyPy <https://www.pypy.org>`__,
`GraalPy <https://www.graalvm.org/python/>`__,
`IronPython <https://ironpython.net>`__,
`RustPython <https://github.com/RustPython/RustPython>`__,
`MicroPython <https://micropython.org>`__,
and `Jython <https://www.jython.org>`__), may take
very different approaches for the implementation of
different subsystems. They need:
```

* Introspect objects, including types, instances, and functions
* Raise and handle exceptions
* Import modules
* Manage threads
* Manage sub-interpreters
* Handle and send signals

The following sections look at the unique requirements of various stakeholders.

Extension Writers
-----------------

Extension writers are the traditional users of the C API. Their requirements
are the universal actions listed above.

Authors of Embedded Python Applications
---------------------------------------

Applications with an embedded Python interpreter. Examples are
`Blender <https://docs.blender.org/api/current/info_overview.html>`__ and
`OBS <https://obsproject.com/wiki/Getting-Started-With-OBS-Scripting>`__.

They need to be able to:

* Configure the interpreter (import paths, inittab, ``sys.argv``, etc.).
* Interact with the execution model and program lifetime, including
  clean interpreter shutdown and restart.
* Represent complex data models in a way Python can use without
  having to create deep copies.
* Provide and import frozen modules.
* Run multiple independent interpreters (in particular, when embedded
  in a library that wants to avoid global effects).

> Alternative Python Implementations

Reviewer 17: Not to harp on the "listing CPython itself as a stakeholder of the C API" topic too much more, but it is very odd that every implementation of Python is listed as a stakeholder, but not CPython itself.

Explicit is better than implicit. :)

Author: Good point. So perhaps we change "Alternative Implementations" to "Implementations" and add cpython itself to the list?

Reviewer: > Good point. So perhaps we change "Alternative Implementations" to "Implementations" and add cpython itself to the list?

It makes sense. I made this exact change.

Reviewer 17: Sounds good.

----------------------------------

Alternative implementations of Python (such as
`PyPy <https://www.pypy.org>`__,
`GraalPy <https://www.graalvm.org/python/>`__,
`IronPython <https://ironpython.net>`__,
`RustPython <https://github.com/RustPython/RustPython>`__,
`MicroPython <https://micropython.org>`__,
and `Jython <https://www.jython.org>`__), may take
very different approaches for the implementation of
different subsystems. They need:

* The API to be abstract and hide implementation details.
* A specification of the API, ideally with a test suite
  that ensures compatibility.
* It would be nice to have an ABI that can be shared
  across Python implementations.

Alternative APIs
----------------

There are several projects that implement alternatives to the

> C API, which offer extension users advantanges over programming

Reviewer 17: Lumping HPy with `pythoncapi-compat ` and `Argument Clinic` is odd, because they are completely different kinds of things. I doubt the authors of `pythoncapi-compat ` and `Argument Clinic` would call them alternatives to the C API -- they're helpers to be used with the existing C API.

It's likewise awkward to call `Cython` a binding generator. It's more like a whole new language.

Could we perhaps create two sections, one with "API Helpers" (with pythoncapi-compat and Argument Clinic) and another for "Alternative APIs" (with HPy and Cython)? I also wouldn't object to each of HPy and Cython being given their own section. They're both significant efforts to reimagine how extension modules are written.

In the HPy section (or somewhere else) could we also link to the HPy [Manifesto](https://github.com/hpyproject/hpy/wiki/c-api-next-level-manifesto) and its [Why](https://github.com/hpyproject/hpy/wiki/c-api-next-level-changes) document as previous works?

Author: See https://github.com/python/peps/pull/3491/files#discussion_r1362383454 re how Cython ended up here (it was previously in the Alternative APIs section).

Reviewer 17: Opening the link doesn't open a specific comment for me, but I did read all of the previous comments and my ones above still stand, I think.

Reviewer 7: Maybe we should just merge "Alternative APIs" and "binding generators"? The needs and concerns are similar, and it's mainly the scope that varies (e.g. Argument Clinic is scoped to parameter conversion, but otherwise it has all the same needs as anything else in this group).

A more interesting distinction could be runtime bindings (e.g. Python.NET) vs. compile-time bindings, but probably that's going to be sufficiently covered by "can't always rely on C macros".

directly with the C API. These APIs are implemented with the
C API, and in some cases by using CPython internals.
Examples are
`HPy <https://hpyproject.org>`__ and
`pythoncapi-compat <https://pythoncapi-compat.readthedocs.io/en/latest/>`__.
CPython's DSL for parsing function arguments, the
`Argument Clinic <https://docs.python.org/3/howto/clinic.html>`__,
can also be seen as belonging to this category of stakeholders.

Such systems need minimal building blocks for accessing CPython
efficiently. They don't necessarily need an ergonomic API, because
they typically generate code that is not intended to be read
by humans. But they do need it to be comprehensive enough so that
they don't need to access internals, without sacrificing performance.

These tools might also benefit from a choice between a more stable
and a more performant (possibly lower-level) API. Their users could
then decide whether they can afford to regenerate the code often or
trade some performance for more stability and less maintenance work.

Binding Generators
------------------

Libraries that create bindings between Python and other object models,
paradigms or languages, such as

> `Cython <https://cython.org>`__,

Reviewer: ```suggestion
`Cython <https://cython.org>`__,
`cffi <https://cffi.readthedocs.io/>`__,
```

`pybind11 <https://pybind11.readthedocs.io/en/stable/>`__ for C++11,
`PyO3 <https://github.com/PyO3/pyo3>`__ for Rust,
`PySide <https://pypi.org/project/PySide/>`__ for Qt,
`PyGObject <https://pygobject.readthedocs.io/en/latest/>`__ for GTK,
`Pygolo <https://gitlab.com/pygolo/py>`__ for Go,
`PyJNIus <https://github.com/kivy/pyjnius/>`__ for Java,
`SWIG <https://swig.org/>`__ for C/C++, or

> `Python.NET <https://github.com/pythonnet/pythonnet>`__ for .NET (C#).

Author: Isn't cffi more like argument clinic than a language binding?

> `Python.NET <https://github.com/pythonnet/pythonnet>`__ for .NET (C#).

Reviewer 3: cffi is more or less like ctypes.

> `Python.NET <https://github.com/pythonnet/pythonnet>`__ for .NET (C#).

Reviewer: > We should probably mention [cffi](https://github.com/python-cffi/cffi) here.

I added cffi to the list. It fits into the definition of this section.

> `Python.NET <https://github.com/pythonnet/pythonnet>`__ for .NET (C#).

Reviewer 17: CFFI can be used in a couple of different ways, but in its main mode it is a means to wrap a C API in a Python C extension -- https://cffi.readthedocs.io/en/stable/overview.html.

Some of the alternative APIs mentioned in the previous section create
alternative object models and therefore they can also be seen as

> belonging to this category.

Reviewer 17: Which of the projects mentioned in "Alternative APIs" creates an alternative object model?

Author: That is indeed obsolete - it was referring to Cython before that was moved to this section. I'll delete this sentence.

They need to:

> * Create custom objects (e.g. function/module objects

Reviewer 17: Perhaps worth a note that many of these needs are relevant to writing extensions in general. The bindings end up having to do more of this, but any sufficiently complex extension might do many of the things listed here.

Author: Wouldn't it be true to any of the API capabilities we list can show up in a sufficiently complex extension module?  Do we want the extension modules section to cover all extensions, or the common, simple extensions?

Reviewer 17: It would be nice to be clear that these are relevant not only to bindings, but also to many large important C extensions.

  and traceback entries) that match the behavior of equivalent
  Python code as closely as possible.
* Dynamically create objects which are static in traditional
  C extensions (e.g. classes/modules), and need CPython to manage
  their state and lifetime.

> * Adapt foreign objects (strings, GC'd containers), with low overhead.

Reviewer 17: Adapting foreign objects with low overhead in particular seems a key goal of many extensions (e.g. numpy).

Reviewer 7: We should emphasise the dynamic nature of what we mean here. These tools _don't know the shape of the object_ until they're running (possibly in a generation step, but certainly run by someone other than the original devs of the tool), whereas an extension like numpy has predefined its own objects.

Author: I'll rephrase to "Dynamically adapt foreign objects ..."

-----------------------------

The difficulty of making changes in the C API is central to this report. It is
implicit in many of the issues we discuss here, particularly when we need to
decide whether an incremental bugfix can resolve the issue, or whether it can
only be addressed as part of an API redesign
[`Issue 44 <https://github.com/capi-workgroup/problems/issues/44>`__]. The
benefit of each incremental change is often viewed as too small to justify the
disruption. Over time, this implies that every mistake we make in an API's
design or implementation remains with us indefinitely.

We can take two views on this issue. One is that this is a problem and the
solution needs to be baked into any new C API we design, in the form of a
process for incremental API evolution. The other possible approach is that
this is not a problem to be solved, but rather a feature of any API. In this
view, API evolution should not be incremental, but rather through large
redesigns, each of which learns from the mistakes of the past and is not
shackled by backwards compatibility requirements. A compromise approach
is somewhere between these two extremes, fixing issues which are easy

> or important enough to tackle incrementally, and leaving others alone.

Reviewer 17: This paragraph presupposes a trade-off which is not necessarily necessary.

With ABI versioning, it is possible for a Python implementation to concurrently support two very different ABIs and so have the best of both worlds. HPy currently implements such a solution.

Maybe one does not like a particular solution, but this isn't a simple zero sum slider.

I don't have a particular edit to suggest, but would we happy if a sentence was added. The paragraph after this one does a great job of describing the crux of the matter -- we needed an agreed upon plan for API evolution.

Author: > With ABI versioning, it is possible for a Python implementation to concurrently support two very different ABIs and so have the best of both worlds. 

This is what you do in both the evolutionary and the revolutionary options. There is no option where you evolve the API/ABI without versioning. The proponents of stability argue that adding new things is fine, but nothing should ever be removed and nobody should ever be forced off the old API. The evolutionary approach is that you do eventually deprecate and remove old version of the API, so that you don't need to maintain them forever.

Author: I changed the text to clarify this.

[`Issue 45 <https://github.com/capi-workgroup/problems/issues/45>`__].

We identified a deeper issue in relation to the way that reference
counting is exposed. The way that C extensions are required to
manage references with calls to ``Py_INCREF`` and ``Py_DECREF`` is
specific to CPython's memory model, and is hard for alternative
Python implementations to emulate.
[`Issue 12 <https://github.com/capi-workgroup/problems/issues/12>`__].

Another set of problems arises from the fact that a ``PyObject*`` is
exposed in the C API as an actual pointer rather than a handle. The
address of an object serves as its ID and is used for comparison,
and this complicates matters for alternative Python implementations
that move objects during GC
[`Issue 37 <https://github.com/capi-workgroup/problems/issues/37>`__].

A separate issue is that object references are opaque to the runtime,
discoverable only through calls to ``tp_traverse``/``tp_clear``,
which have their own purposes. If there was a way for the runtime to
know the structure of the object graph, and keep up with changes in it,
this would make it possible for alternative implementations to implement
different memory management schemes

> [`Issue 33 <https://github.com/capi-workgroup/problems/issues/33>`__].

Reviewer 17: Could we add a short paragraph about defining the ABI too? Currently the ABI definition is also implicit whether having a much more explicit and simpler ABI definition would help seems an important question for future discussion.

Author: Can you add an issue on the problem repo that I can link to?

Reviewer 17: The most relevant issue seems to be this discussion of the ABI -- https://github.com/capi-workgroup/problems/issues/39.

At the moment Python's "Stable ABI" is seen as some subset of the API, but really these are two completely different kinds of things that might need quite different approaches. The API is a C interface, but the ABI is compiled library binary interface.

Author: I don't think this issue mentions the lack of clear specification. But I think I can just add a sentence to the preamble of this section, without referring to an issue.

Reviewer 17: The issue is more of a direct demonstration of the lack.

Reviewer 18: > At the moment Python's "Stable ABI" is seen as some subset of the API, but really these are two completely different kinds of things that might need quite different approaches.

Is it? It's very explicitly called the Limited API and using only the limited API allows safely relying on the stable ABI.

Reviewer 17: Yes, this is the misconception of what an ABI is that I was referring to.

Currently in CPython the C ABI is defined via the Limit API, but it doesn't have to be this way. For example, in HPy the API and ABI are rather different. The HPy API looks a lot like the old C API, but the ABI consists almost entirely of the context object (plus a few functions for determining the context version and initializing the extension).

PEP: 733
Title: An Evaluation of Python's Public C API
Author: Erlend Egeberg Aasland <erlend@python.org>,
        Domenico Andreoli <domenico.andreoli@linux.com>,
        Stefan Behnel <stefan_ml@behnel.de>,
        Carl Friedrich Bolz-Tereick <cfbolz@gmx.de>,
        Simon Cross <hodgestar@gmail.com>,
        Steve Dower <steve.dower@python.org>,
        Tim Felgentreff <tim.felgentreff@oracle.com>,
        David Hewitt <1939362+davidhewitt@users.noreply.github.com>,
        Shantanu Jain <hauntsaninja at gmail.com>,
        Wenzel Jakob <wenzel.jakob@epfl.ch>,
        Irit Katriel <irit@python.org>,
        Marc-Andre Lemburg <mal@lemburg.com>,
        Donghee Na <donghee.na@python.org>,
        Karl Nelson <nelson85@llnl.gov>,
        Ronald Oussoren <ronaldoussoren@mac.com>,
        Antoine Pitrou <solipsis@pitrou.net>,
        Neil Schemenauer <nas@arctrix.com>,
        Mark Shannon <mark@hotpy.org>,
        Stepan Sindelar <stepan.sindelar@oracle.com>,
        Gregory P. Smith <greg@krypto.org>,
        Eric Snow <ericsnowcurrently@gmail.com>,
        Victor Stinner <vstinner@python.org>,

>         Serhiy Storchaka <storchaka@gmail.com>,
>         Petr Viktorin <encukou@gmail.com>,

Reviewer 19: ```suggestion
        Serhiy Storchaka <storchaka@gmail.com>,
        Guido van Rossum <guido@python.org>,
        Petr Viktorin <encukou@gmail.com>,
```

C API, and in some cases by using CPython internals.
Examples are
`HPy <https://hpyproject.org>`__ and
`pythoncapi-compat <https://pythoncapi-compat.readthedocs.io/en/latest/>`__.
CPython's DSL for parsing function arguments, the
`Argument Clinic <https://devguide.python.org/development-tools/clinic/>`__,
can also be seen as belonging to this category of stakeholders.

Such systems need minimal building blocks for accessing CPython
efficiently. They don't necessarily need an ergonomic API, because
they typically generate code that is not intended to be read
by humans. But they do need it to be comprehensive enough so that
they don't need to access internals, without sacrificing performance.

These tools might also benefit from a choice between a more stable
and a more performant (possibly lower-level) API. Their users could
then decide whether they can afford to regenerate the code often or
trade some performance for more stability and less maintenance work.

Binding Generators
------------------

Libraries that create bindings between Python and other object models,
paradigms or languages, such as
`Cython <https://cython.org>`__,
`cffi <https://cffi.readthedocs.io/>`__,

> `pybind11 <https://pybind11.readthedocs.io/en/stable/>`__ for C++11,

Reviewer 20: ```suggestion
`pybind11 <https://pybind11.readthedocs.io/en/stable/>`__ for C++11,
`nanobind <https://github.com/wjakob/nanobind>`__ for C++11,
```

Reviewer 7: Perhaps combine the two into a single line: "pybind11 and nanobind for C++11"? The links will render as a single word, and this list is only examples, not exhaustive.

Reviewer 3: nanobind requires C++17, so no.

Reviewer 7: Perhaps just "for C++" then? Again, examples, not a table of all the bindings that exist.

PEP: 733
Title: An Evaluation of Python's Public C API

> Author: Erlend E. Aasland <erlend@python.org>,

Reviewer 21: ```suggestion
Author: Erlend Egeberg Aasland <erlend@python.org>,
```

CPython's DSL for parsing function arguments, the
`Argument Clinic <https://devguide.python.org/development-tools/clinic/>`__,
can also be seen as belonging to this category of stakeholders.

Alternative APIs need minimal building blocks for accessing CPython
efficiently. They don't necessarily need an ergonomic API, because
they typically generate code that is not intended to be read
by humans. But they do need it to be comprehensive enough so that
they can avoid accessing internals, without sacrificing performance.

Binding generators often need to:

* Create custom objects (e.g. function/module objects
  and traceback entries) that match the behavior of equivalent
  Python code as closely as possible.
* Dynamically create objects which are static in traditional
  C extensions (e.g. classes/modules), and need CPython to manage
  their state and lifetime.
* Dynamically adapt foreign objects (strings, GC'd containers), with
  low overhead.
* Adapt external mechanisms, execution models and guarantees to the
  Python way (green threads/continuations, one-writer-or-multiple-readers
  semantics, virtual multiple inheritance, 1-based indexing, super-long

>   inheritance chains, goroutines, channels, etc.).

Reviewer 22: If by "fiber" you mean a stackful coroutine, then that's greenlets. So it is a fairly important use case.
Green threads are almost the same, but imply the existence of a scheduler.
Maybe replace `green threads/continuations, ...` with `stackful coroutines, continuations, ...`

These tools might also benefit from a choice between a more stable

> and a more performant (possibly lower-level) API. Their users could

Reviewer 22: According to bing "performant" means "functioning well or as expected".
Maybe change "more performant" to "lower overhead" or just "faster"

trade some performance for more stability and less maintenance work.

Strengths of the C API
======================

While the bulk of this document is devoted to problems with the
C API that we would like to see fixed in any new design, it is
also important to point out the strengths of the C API, and to
make sure that they are preserved.

As mentioned in the introduction, the C API enabled the
development and growth of the Python ecosystem over the last
three decades, while evolving to support use cases that it was
not originally designed for. This track record in itself is
indication of how effective and valuable it has been.

A number of specific strengths were mentioned in the
capi-workgroup discussions. Heap types were identified
as much safer and easier to use than static types
[`Issue 4 <https://github.com/capi-workgroup/problems/issues/4#issuecomment-1542324451>`__].

API functions that take a C string literal for lookups based
on a Python string are very convenient
[`Issue 30 <https://github.com/capi-workgroup/problems/issues/30#issuecomment-1550098113>`__].

The Limited API and stable ABI hide implementation details and
make it easier to evolve Python

> [`Issue 30 <https://github.com/capi-workgroup/problems/issues/30#issuecomment-1560083258>`__].

Reviewer 22: Any API that hides implementation details makes it easier to evolve the implementation. It doesn't have to be limited.
If the point here is that a limited API helps evolution, I disagree.
An API that is not tied to implementation details is what is important. It does have to be "limited".

So maybe change this to something like:
"The limited API shows that an API that hide implementation details makes it easier to evolve Python"
The stability of the ABI and the extent of the API are also largely unrelated.

C API problems
==============

The remainder of this document summarizes and categorizes the problems that were reported on
the `capi-workgroup <https://github.com/capi-workgroup/problems/issues/>`__ repository.
The issues are grouped into several categories.

API Evolution and Maintenance
-----------------------------

The difficulty of making changes in the C API is central to this report. It is
implicit in many of the issues we discuss here, particularly when we need to
decide whether an incremental bugfix can resolve the issue, or whether it can
only be addressed as part of an API redesign
[`Issue 44 <https://github.com/capi-workgroup/problems/issues/44>`__]. The
benefit of each incremental change is often viewed as too small to justify the
disruption. Over time, this implies that every mistake we make in an API's

> design or implementation remains with us indefinitely.

Author: Too cliche, I think.

manage references with calls to ``Py_INCREF`` and ``Py_DECREF`` is
specific to CPython's memory model, and is hard for alternative
Python implementations to emulate.
[`Issue 12 <https://github.com/capi-workgroup/problems/issues/12>`__].

Another set of problems arises from the fact that a ``PyObject*`` is
exposed in the C API as an actual pointer rather than a handle. The
address of an object serves as its ID and is used for comparison,
and this complicates matters for alternative Python implementations
that move objects during GC
[`Issue 37 <https://github.com/capi-workgroup/problems/issues/37>`__].

A separate issue is that object references are opaque to the runtime,
discoverable only through calls to ``tp_traverse``/``tp_clear``,
which have their own purposes. If there was a way for the runtime to
know the structure of the object graph, and keep up with changes in it,
this would make it possible for alternative implementations to implement
different memory management schemes
[`Issue 33 <https://github.com/capi-workgroup/problems/issues/33>`__].

Object Reference Management
---------------------------

There are C API functions that return borrowed references, and

> functions that steal references to arguments, but there isn't a

Reviewer 22: I still don't like the term steal. It is unnecessarily pejorative. 
"Stealing" a reference is just a transfer of ownership from caller to callee, which has many legitimate use cases,
particularly when the caller no longer needs the reference. "Stealing" reduces the number of decrefs and reduces the lifetime of unwanted objects.
Rust calls such a transfer of ownership a "move", but that is ownership of the object, not the reference.

If you really must use "steal", then please add a note explaining the term.

Reviewer 7: Maybe for the purposes of this document, simply writing `"steal"` to acknowledge it's the commonly used term and doesn't necessarily have its usual implications is sufficient?

If we don't have a linkable glossary entry already, this isn't the place to write the first definition. (We probably have a linkable glossary entry though, so maybe quotes and a link to that will prevent people from being upset if this is where they first encounter the term.)

Author: I'll just say "decref arguments".

naming convention that makes this obvious, so this is error prone
[`Issue 8 <https://github.com/capi-workgroup/problems/issues/8>`__
and `Issue 52 <https://github.com/capi-workgroup/problems/issues/52>`__].
The terminology used to describe these situations in the documentation
can also be improved
[`Issue 11 <https://github.com/capi-workgroup/problems/issues/11>`__].

A more radical change is necessary in the case of functions that

> return borrowed references (such as ``PyList_GetItem``)

Reviewer 22: Can we add a section somewhere describing our reference terminology. We are using the same name for different things are it should be made clear.

The term "borrowed" has two different meanings for return values and for arguments.
For arguments, the caller loans a reference to the callee. Perfectly good and safe.
When "returning a borrowed reference", we don't really have a reference to an object at all, just a pointer to memory which we hope is a valid object.
I'd suggest "volatile reference", but this probably isn't the place to be coining new terms.

Reviewer 7: As with steal, maybe `"borrowed"` is an easy way here to use the term we usually use, even though it doesn't really apply?

This is definitely not the place to coin new terms. Plus "volatile" carries as much baggage as "borrowed" (I'm sure it's only a matter of time before someone decides that "volatile" is only allowed to describe unstable people (can I still use "unstable" here?) and we can't use it for technical reasons anymore...), so would also need just as clear a definition.

I'm sure one of the issues somewhere is about clearly defining the terms. That's an important future discussion, but this document should reflect that we currently refer to them as "borrowed".

Author: I can just write explicitly:

```
A more radical change is necessary in the case of a function that
returns a ``PyObject*`` that it did not incref (such as ``PyList_GetItem``)
```

Reviewer 22: Incrementing a reference count is an implementation detail. The problem is lack of ownership of the reference.
Can we put "borrow" and "steal" in quotes as @zooba suggests.
We do need a definition somewhere though.

Here's some definitions:
A "stolen" reference is one where the ownership is transferred from caller to callee at the call, permanently.
A "borrowed" parameter reference is where the reference is loaned to the callee by the caller for exactly the duration of the call.
A "borrowed" return reference is not an owned reference at all, just a pointer. We just hope that there exists a reference to the object that will outlive the pointer.

Author: I reworded this section. How's this version?

In both cases, the reference/pointer is valid for as long as the
owning object holds the reference, but this time is hard to reason about.
Such functions should not exist in the API without a mechanism that can
make them safe.

For containers, the API is currently missing bulk operations on the
references of contained objects. This is particularly important for
a stable ABI where ``INCREF`` and ``DECREF`` cannot be macros, making
bulk operations expensive when implemented as a sequence of function
calls
[`Issue 15 <https://github.com/capi-workgroup/problems/issues/15>`__].

Type Definition and Object Creation
-----------------------------------

The C API has functions that make it possible to create incomplete
or inconsistent Python objects, such as ``PyTuple_New`` and
``PyUnicode_New``. This causes problem when the object is tracked
by GC or its ``tp_traverse``/``tp_clear`` functions are called.
Such functions should be removed from the C API. Related functions,
such as ``PyTuple_SetItem`` which is used to modify a partially
initialized tuple, should also be removed (tuples are immutable
once fully initialized)

> [`Issue 56 <https://github.com/capi-workgroup/problems/issues/56>`__].

Reviewer 23: I expect this to be contentious for some users. Just because a tuple is immutable **at the Python level** doesn't mean all of its contents need exist up front elsewhere in a form that an acceptable non-partially-formed object API building a tuple could use. The problem statement here is real, but the solution statement of "just get rid of these APIs, effectively telling people who want efficient tuple creation they're SOL" doesn't smell right.

I'm not claiming that we'll do any differently. But that we should expect objections from real users if we do so.

I see there is more useful active discussion on this in the linked issue (great!).  so just take this comment as an FYI about how I expect some readers to react...

tiers are specified and consider designing a method that will unify the
way we currently select between the different tiers
[`Issue 59 <https://github.com/capi-workgroup/problems/issues/59>`__].

API elements whose names begin with an underscore are considered
private, essentially an API tier with no stability guarantees.
However, this was only clarified recently, in :pep:`689`. It is
not clear what the change policy should be with respect to such
API elements that predate PEP 689
[`Issue 58 <https://github.com/capi-workgroup/problems/issues/58>`__].

There are API functions which have an unsafe (but fast) version as well as
a safe version which performs error checking (for example,
``PyTuple_GET_ITEM`` vs ``PyTuple_GetItem``). It may help to
be able to group them into their own tiers - the "unsafe API" tier and
the "safe API" tier
[`Issue 61 <https://github.com/capi-workgroup/problems/issues/61>`__].

Use of the C Language
---------------------

A number of issues were raised with respect to the way that CPython
uses the C language. First there is the issue of which C dialect
we use, and how we test our compatibility with it

> [`Issue 42 <https://github.com/capi-workgroup/problems/issues/42>`__].

Reviewer 23: perhaps add "as well as API header compatibility with C++ dialects" to this as that is important and seems to be covered by that issue as well.

identified a number of recurring themes. The sections below
mostly correspond to these themes, and each contains a combined
description of the issues raised in that category, along with
links to the individual issues. In addition, we included a section
that aims to identify the different stakeholders of the C API,
and the particular requirements that each of them has.

C API Stakeholders
==================

As mentioned in the introduction, the C API was originally
created as the internal interface between CPython's
interpreter and the Python layer. It was later exposed as
a way for third-party developers to extend and embed Python
programs. Over the years, new types of stakeholders emerged,
with different requirements and areas of focus. This section
describes this complex state of affairs in terms of the
actions that different stakeholders need to perform through
the C API.

Common Actions for All Stakeholders
-----------------------------------

There are actions which are generic, and required by
all types of API users:

* Define functions and call them
* Define new types
* Create instances of builtin and user-defined types

>   and perform operations on them

Reviewer 24: I was referring to C APIs which work on objects which the interpreter (e.g. sys.path) or other extensions created. Your list currently only includes and permits operations on objects created by the extension itself. That's too narrow in scope.

Regarding the reference counting scheme, perhaps it's better to include "* Access to memory management".

>   and perform operations on them

Author: I see. I'll split the sentence into two so this implication would not be there:

```
* Create instances of builtin and user-defined types
* Perform operations on object instances
```

`OBS <https://obsproject.com/wiki/Getting-Started-With-OBS-Scripting>`__.

They need to be able to:

* Configure the interpreter (import paths, inittab, ``sys.argv``, etc.).
* Interact with the execution model and program lifetime, including
  clean interpreter shutdown and restart.
* Represent complex data models in a way Python can use without
  having to create deep copies.
* Provide and import frozen modules.
* Run and manage multiple independent interpreters (in particular, when
  embedded in a library that wants to avoid global effects).

Python Implementations
----------------------

Python implementations such as
`CPython <https://www.python.org>`__,
`PyPy <https://www.pypy.org>`__,
`GraalPy <https://www.graalvm.org/python/>`__,
`IronPython <https://ironpython.net>`__,
`RustPython <https://github.com/RustPython/RustPython>`__,
`MicroPython <https://micropython.org>`__,
and `Jython <https://www.jython.org>`__), may take
very different approaches for the implementation of
different subsystems. They need:

* The API to be abstract and hide implementation details.
* A specification of the API, ideally with a test suite

>   that ensures compatibility.

Reviewer 7: Agreed, but it's not important to make it at this stage in this document. It's definitely been discussed for future designs, though (in the capi-workgroup repositories).

>   that ensures compatibility.

Author: This document is about the Python C API. Not the CPython C API.

>   that ensures compatibility.

Reviewer 24: Perhaps you should make that clear somewhere near the start of the document, since traditionally, the "Python C API" is being identified with whatever the CPython implementation exports (along with the different tiers that have popped up over the years).

* It would be nice to have an ABI that can be shared
  across Python implementations.

Alternative APIs and Binding Generators
---------------------------------------

There are several projects that implement alternatives to the
C API, which offer extension users advantanges over programming
directly with the C API. These APIs are implemented with the
C API, and in some cases by using CPython internals.

There are also libraries that create bindings between Python and
other object models, paradigms or languages.

There is overlap between these categories: binding generators
usually provide alternative APIs, and vice versa.

Examples are
`Cython <https://cython.org>`__,
`cffi <https://cffi.readthedocs.io/>`__,
`pybind11 <https://pybind11.readthedocs.io/en/stable/>`__ and
`nanobind <https://github.com/wjakob/nanobind>`__ for C++,
`PyO3 <https://github.com/PyO3/pyo3>`__ for Rust,
`PySide <https://pypi.org/project/PySide/>`__ for Qt,
`PyGObject <https://pygobject.readthedocs.io/en/latest/>`__ for GTK,
`Pygolo <https://gitlab.com/pygolo/py>`__ for Go,

> `PyJNIus <https://github.com/kivy/pyjnius/>`__ for Java,

Reviewer 25: It is awkward to ask someone to review a request and then endorse a competitor.   The contribution to the Python CAPI review on which portions of this PEP was based was performed Karl Nelson (aka Thrameos) who write for JPype which is Java binding for Python for window/unix not PyJNIus which is the Java binding for Andriod.

Reviewer 25: Beyond this minor quip, I have reviewed the much of this PEP and whole hardly agree with the goal and the accuracy of the stated problems. I would plus up many of the specific subsections as problems that i frequently have encountered if it didn't clutter up the review process.

Reviewer 7: Nothing in the PEP constitutes an "endorsement" of any of the examples listed here. This is not a list of recommendations, just an illustration that there's a thriving ecosystem of tools in this category (which is important, because most people wouldn't be able to name a single one, let alone recognise that it's an important type of project for us to support).

Reviewer 5: Seems worth to also mention [JPype](https://jpype.readthedocs.io/en/latest/) though. I'll send a suggestion for that.

> `Pygolo <https://gitlab.com/pygolo/py>`__ for Go,
> `PyJNIus <https://github.com/kivy/pyjnius/>`__ for Java,

Reviewer 5: ```suggestion
`Pygolo <https://gitlab.com/pygolo/py>`__ for Go,
`JPype <https://github.com/jpype-project/jpype/>`__ and
`PyJNIus <https://github.com/kivy/pyjnius/>`__ for Java,
```

shackled by backwards compatibility requirements (in the meantime, new
API elements may be added, but nothing can ever be removed). A compromise
approach is somewhere between these two extremes, fixing issues which are
easy or important enough to tackle incrementally, and leaving others alone.

The problem we have in CPython is that we don't have an agreed, official
approach to API evolution. Different members of the core team are pulling in
different directions and this is an ongoing source of disagreements.
Any new C API needs to come with a clear decision about the model
that its maintenance will follow, as well as the technical and
organizational processes by which this will work.

If the model does include provisions for incremental evolution of the API,
it will include processes for managing the impact of the change on users
[`Issue 60 <https://github.com/capi-workgroup/problems/issues/60>`__],
perhaps through introducing an external backwards compatibility module
[`Issue 62 <https://github.com/capi-workgroup/problems/issues/62>`__],
or a new API tier of "blessed" functions
[`Issue 55 <https://github.com/capi-workgroup/problems/issues/55>`__].

> 

Reviewer 7: This is also being discussed for future plans, but doesn't need to be captured in this document.

> 

Author: Agree with @zooba. Also, I think the section on "API Evolution and Maintenance" covers many of these questions.

once, so if the arg is an expression with side effects, they are
duplicated
[`Issue 3 <https://github.com/capi-workgroup/problems/issues/3>`__].

The meaning of ``Py_SIZE`` depends on the type and is not always
reliable
[`Issue 10 <https://github.com/capi-workgroup/problems/issues/10>`__].

Some API function do not have the same behaviour as their Python
equivalents.  The behaviour of ``PyIter_Next`` is different from
``tp_iternext``.
[`Issue 29 <https://github.com/capi-workgroup/problems/issues/29>`__].
The behaviour of ``PySet_Contains`` is different from ``set.__contains__``
[`Issue 6 <https://github.com/capi-workgroup/problems/issues/6>`__].

The fact that ``PyArg_ParseTupleAndKeywords`` takes a non-const
``char*`` array as argument makes it more difficult to use
[`Issue 28 <https://github.com/capi-workgroup/problems/issues/28>`__].

``Python.h`` does not expose the whole API. Some headers (like ``marshal.h``)
are not included from ``Python.h``.
[`Issue 43 <https://github.com/capi-workgroup/problems/issues/43>`__].

**Naming**

``PyLong`` and ``PyUnicode`` use names which no longer match the Python
types they represent (``int``/``str``). This can be fixed in a new API

> [`Issue 14 <https://github.com/capi-workgroup/problems/issues/14>`__].

Reviewer 24: I'd write "could be fixed", since it's not clear whether the added churn is really beneficial.

This section consists of a list of feature requests, i.e., functionality
that was identified as missing in the current C API.

Debug Mode
~~~~~~~~~~

A debug mode that can be activated without recompilation and which
activates various checks that can help detect various types of errors
[`Issue 36 <https://github.com/capi-workgroup/problems/issues/36>`__].

Introspection
~~~~~~~~~~~~~

There aren't currently reliable introspection capabilities for objects
defined in C in the same way as there are for Python objects
[`Issue 32 <https://github.com/capi-workgroup/problems/issues/32>`__].

Efficient type checking for heap types
[`Issue 17 <https://github.com/capi-workgroup/problems/issues/17>`__].

Improved Interaction with Other Languages
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Interfacing with other GC based languages, and integrating their
GC with Python's GC
[`Issue 19 <https://github.com/capi-workgroup/problems/issues/19>`__].

Inject foreign stack frames to the traceback
[`Issue 18 <https://github.com/capi-workgroup/problems/issues/18>`__].

Concrete strings that can be used in other languages
[`Issue 16 <https://github.com/capi-workgroup/problems/issues/16>`__].

> 

Reviewer 24: It may be useful to add a section "Future Developments" which mentions things that will in one form or another affect the future C API and are on the roadmap, e.g. the no-GIL patches, or possible future changes such as moving away from reference counting, JIT compilation, sharing Python objects between different in-process interpreter instances.

Reviewer 7: Mentioning that no-GIL is in progress against the current API seems like helpful context to capture here.

More speculative changes will get their own discussions and PEPs. There's more than enough speculation earlier in the doc already...

Author: We could mention this in the intro: where we talk about the history, say that there will be more developments coming.

Reviewer 24: Agreed and I believe that's a good idea in general to not make it look like this document will be the only one defining the future of the Python C API.

The problem we have in CPython is that we don't have an agreed, official
approach to API evolution. Different members of the core team are pulling in
different directions and this is an ongoing source of disagreements.
Any new C API needs to come with a clear decision about the model
that its maintenance will follow, as well as the technical and
organizational processes by which this will work.

If the model does include provisions for incremental evolution of the API,
it will include processes for managing the impact of the change on users
[`Issue 60 <https://github.com/capi-workgroup/problems/issues/60>`__],
perhaps through introducing an external backwards compatibility module
[`Issue 62 <https://github.com/capi-workgroup/problems/issues/62>`__],
or a new API tier of "blessed" functions
[`Issue 55 <https://github.com/capi-workgroup/problems/issues/55>`__].

API Specification and Abstraction
---------------------------------

The C API does not have a formal specification, it is described
semi-formally in the documentation and exposed through C header
files. This is also true for the limited API and the stable ABI.

> This creates a number of problems.

Reviewer 16: This is a nitpick, but the limited API and stable ABI *do* have a machine-readable representation for external tools to consume: <https://github.com/python/cpython/blob/main/Misc/stable_abi.toml>. It might be worth pointing that out, as a starting point or source of inspiration for improving the overall specify-ability of the C API.

(I use that representation to generate `abi3info`, which in turn powers tools like `abi3audit`.)

Author: Since we're focused here on problems rather than solutions, I think this could be worth mentioning on the relevant issue, but it wouldn't add clarity to this section (which currently doesn't mention the stable ABI or limited API at all).

Reviewer 16: The point about solutions is a fair one 🙂 

>  but it wouldn't add clarity to this section (which currently doesn't mention the stable ABI or limited API at all).

Forgive me if I'm misunderstanding, but is this true? Here's the part of the paragram that I honed in on:

> This is also true for the limited API and the stable ABI.

My interpretation of that sentence is that it extends the claim about the lack of a formal specification to the limited API and stable ABI. But perhaps that isn't what was intended?

Author: Ah right, sorry. Let me think about how this could be corrected.

Reviewer 7: "Formal specification" isn't the same as "machine readable representation". The header files are technically machine readable representations, but that doesn't mean the whole interpreter is formally specified 😉 

The specification is currently the reference interpreter, when it should be a written document full of "SHOULDs" and "SHALLs" and stuff like that. How about something like:

"The C API does not have a formal specification, but is currently defined as whatever the reference implementation (CPython) defines in a particular version. The documentation acts as an incomplete description, but is not sufficient for verifying the correctness of either the full API, the limited API, or the stable ABI. As a result, the C API may change significantly between releases without needing a more visible specification update, which leads to a number of problems." (I assume you go on to describe these problems, otherwise the last bit is a bit strange, but I don't want to do the seven clicks to get GitHub to expand it without losing my comment right now.)

Reviewer 16: > "Formal specification" isn't the same as "machine readable representation". The header files are technically machine readable representations, but that doesn't mean the whole interpreter is formally specified 😉

Yep, this is why I called it a nitpick and said "machine representation." I thought it would possibly be worth noting that there *are* other (very limited!) descriptions of the C API besides the source code itself.

But anybody who is so motivated will be able to discover those anyways, so I don't feel strongly about this 🙂

Efficient type checking for heap types
[`Issue 17 <https://github.com/capi-workgroup/problems/issues/17>`__].

Improved Interaction with Other Languages
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Interfacing with other GC based languages, and integrating their
GC with Python's GC
[`Issue 19 <https://github.com/capi-workgroup/problems/issues/19>`__].

Inject foreign stack frames to the traceback
[`Issue 18 <https://github.com/capi-workgroup/problems/issues/18>`__].

Concrete strings that can be used in other languages
[`Issue 16 <https://github.com/capi-workgroup/problems/issues/16>`__].

References
==========

1. `Python/C API Reference Manual <https://docs.python.org/3/c-api/index.html>`__
2. `2023 Language Summit Blog Post: Three Talks on the C API <https://pyfound.blogspot.com/2023/05/the-python-language-summit-2023-three.html>`__
3. `capi-workgroup on GitHub <https://github.com/capi-workgroup>`__
4. `Irit's Core Sprint 2023 slides about C API workgroup <https://github.com/iritkatriel/talks/blob/main/2023_Sprint_Brno_C_API.pdf>`__
5. `Petr's Core Sprint 2023 slides <https://drive.google.com/file/d/148NLRPXGZGI1SXfKLMFvQc_iv67hPJQS/view?usp=sharing>`__

> 6. 🔒 `Steve's Discord post with Core Sprint 2023 slides for Things to Learn from HPy <https://discord.com/channels/854719841091715092/897862918115315813/1160942893696024578>`__

Reviewer: Thanks for publishing your slides :-)
